### ISSUE TITLE
[clang] Poor diagnostic with wrong template argument type

### ISSUE THREAD

**Issue by antoniofrighetto** [llvm/llvm-project#121235]
Clang diagnostic for invalid bool-to-enum template parameter conversion lacks failure detail when compared to GCC, as the following C++ snippet:

```cpp
enum endianness { little = 0, big };

template <endianness e>
inline void foo() {}

void bar() {
    foo<false>();
}
```
Fails as follows on latest clang:
```
<source>:7:5: error: no matching function for call to 'foo'
    7 |     foo<false>();
      |     ^~~~~~~~~~
<source>:4:13: note: candidate template ignored: invalid explicitly-specified argument for template parameter 'big_endian'
    4 | inline void foo() {}
```

GCC explicitly indicates the conversion failure instead:
```
<source>:4:13: note: template argument deduction/substitution failed:
<source>:7:15: error: could not convert 'false' from 'bool' to 'endianness'
    7 |     foo<false>();
      |     ~~~~~~~~~~^~
      |               |
      |               bool
```

Godbolt: https://godbolt.org/z/hnazTn3M7 (cc/ @shafik).

**Comment by shafik** [comment]
CC @erichkeane

**Comment by erichkeane** [comment]
I agree, we could definitely do a much better job here.  I suspect our template diagnostics would need an additional 'type' for the note.  It isn't quite a good first issue, but this shouldn't be too awful.  I'll mark as 'good first issue', though note this'll be one of the harder ones

**Comment by AidanGoldfarb** [comment]
I would be happy to pick this up and give it a shot

**Comment by AidanGoldfarb** [comment]
I am making progress and am gaining familiarity with the code. I found that `note_ovl_candidate_explicit_arg_mismatch_named` is what controls this message. My current idea is to modify [SemaOverload](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaOverload.cpp#L11698) from

```c++
S.Diag(Templated->getLocation(),
             diag::note_ovl_candidate_explicit_arg_mismatch_named)
          << ParamD->getDeclName();
```

to something like

```c++
auto Param = *DeductionFailure.getFirstArg(); //not valid
auto Arg = *DeductionFailure.getSecondArg(); //not valid 
S.Diag(Templated->getLocation(),
      diag::note_ovl_candidate_explicit_arg_mismatch_named)
    << ParamD->getDeclName()
    << Param << Arg;
```

Although at the moment the dereferencing of `Param` and `Arg` is not valid (segfault). Due to [getFirstArg and getSecondArg](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaOverload.cpp#L941) directly returning `nullptr` in the case of `InvalidExplicitArguments`. 

- - -

I am currently trying to work backwards from this to discover the control flow. Stepping forwards with a debugger has not been working out, but I think I will eventually figure it out. If this is not the right direction please let me know, or if there are any tips or pointers I would greatly appreciate them!

**Comment by zyn0217** [comment]
This is fixed in clang 23.
